### PR TITLE
Update map generation

### PR DESCRIPTION
**Pull request by tylervipond**

### CODE REVIEW

File: src/spawner.rs [hunk 1]
```diff
@@ -281,7 +289,7 @@ pub fn spawn_item_entities_for_room(ecs: &mut World, rect: &Rect, level: i32) {
   }
 }
 
-pub fn spawn_entities_for_room(ecs: &mut World, rect: &Rect, level: i32) {
+pub fn spawn_entities_for_room(ecs: &mut World, rect: &Rect, level: &Level) {
   spawn_monster_entities_for_room(ecs, rect, level);
   spawn_item_entities_for_room(ecs, rect, level);
 }
```
Comment (alexeagleson) [inline src/spawner.rs:295]: What is the significance of the change of the i32 type to &Level?

Is it that level previously was just a generic integer for keeping track of the floor, and now it's a Struct that carries more information?  And this function now "borrows" access to it for the duration of the function?

File: src/dungeon/operations.rs [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+    let stairs_position = level.rooms[0].center();
+    let stairs_idx = xy_idx(level, stairs_position.0, stairs_position.1);
+    level.tiles[stairs_idx as usize] = TileType::UpStairs;
+    level.stairs_up = Some(Point::new(stairs_position.0, stairs_position.1));
+}
+
+pub fn entities_at_xy(level: &Level, x: i32, y: i32) -> Vec<Entity> {
+    let idx = xy_idx(level, x, y);
+    level.tile_content[idx as usize].to_vec()
+}
```
Comment (alexeagleson) [inline src/dungeon/operations.rs:72]: Is the `entities_at_xy` function meant to be used as a means to identify all obstacles/objects at a particular location?  Is it exclusive to Entity type and unrelated to checking for a "wall" TileType for example?

File: src/systems/map_indexing_system.rs [hunk 3]
```diff
@@ -21,17 +24,17 @@ impl<'a> System<'a> for MapIndexingSystem {
...
         let _p = blockers.get(entity);
         if let Some(_p) = _p {
-          map.blocked[idx as usize] = true
+          level.blocked[idx] = true
         }
-        map.tile_content[idx as usize].push(entity);
+        level.tile_content[idx].push(entity);
       }
     }
   }
```
Comment (alexeagleson) [inline src/systems/map_indexing_system.rs:40]: Is this system in change of keeping track of what entities exist within a particular map/level?  If so what is the significance / benefit of this compared to doing a .join() on all the entities that match a position or location on a level? Is it to give you a way of quickly grabbing a list of entities at a location when you have an (x, y) value?

File: src/dungeon/operations.rs (hunk 2 above)
Comment (tylervipond) [reply src/dungeon/operations.rs:72]: that's correct, tile_content holds a Vec of all entities at a given tile on a map, it is updated every time an entity moves.

File: src/systems/map_indexing_system.rs (hunk 3 above)
Comment (tylervipond) [reply src/systems/map_indexing_system.rs:40]: That's exactly right. If we wanted to use a join to get this information we would have to look up all entities with `Position` and `DungeonLevel` (which would be every entity in the game)  and filter by dungeon level and position, this provides a high speed way of querying a specific tile.

File: src/spawner.rs (hunk 1 above)
Comment (tylervipond) [reply src/spawner.rs:295]: So, the main thing here is that all of the map functions have been moved to out of the impl, and now take a Level object to operate on. That means that on line 28 if we want to get the x,y of a given index we need to also have a level to check that index against.

**Pull request merged**